pkg/sink/codec/bootstraper.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codec
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/model"
+	"go.uber.org/zap"
+)
+
+const (
+	bootstrapWorkerTickerInterval = 5 * time.Second
+	bootstrapWorkerGCInterval     = 30 * time.Second
+
+	defaultMaxInactiveDuration = 30 * time.Minute
+)
+
+type bootstrapWorker struct {
+	changefeedID            model.ChangeFeedID
+	activeTables            sync.Map
+	builder                 RowEventEncoderBuilder
+	sendBootstrapInterval   time.Duration
+	sendBootstrapInMsgCount int32
+	// maxInactiveDuration is the max duration that a table can be inactive
+	maxInactiveDuration time.Duration
+	outCh               chan<- *future
+}
+
+// newBootstrapWorker creates a new bootstrapGenerator instance
+func newBootstrapWorker(
+	changefeedID model.ChangeFeedID,
+	outCh chan<- *future,
+	builder RowEventEncoderBuilder,
+	sendBootstrapInterval time.Duration,
+	sendBootstrapInMsgCount int32,
+	maxInactiveDuration time.Duration,
+) *bootstrapWorker {
+	return &bootstrapWorker{
+		changefeedID:            changefeedID,
+		outCh:                   outCh,
+		builder:                 builder,
+		activeTables:            sync.Map{},
+		sendBootstrapInterval:   sendBootstrapInterval,
+		sendBootstrapInMsgCount: sendBootstrapInMsgCount,
+		maxInactiveDuration:     maxInactiveDuration,
+	}
+}
+
+func (b *bootstrapWorker) run(ctx context.Context) error {
+	sendTicker := time.NewTicker(bootstrapWorkerTickerInterval)
+	defer sendTicker.Stop()
+	gcTicker := time.NewTicker(bootstrapWorkerGCInterval)
+	defer gcTicker.Stop()
+	errCh := make(chan error, 1)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-sendTicker.C:
+			b.activeTables.Range(func(key, value interface{}) bool {
+				table := value.(*tableStatus)
+				err := b.sendBootstrapMsg(ctx, table)
+				if err != nil {
+					errCh <- err
+					return false
+				}
+				return true
+			})
+		case <-gcTicker.C:
+			b.gcInactiveTables()
+		case err := <-errCh:
+			return err
+		}
+	}
+}
+
+func (b *bootstrapWorker) addEvent(
+	ctx context.Context,
+	key TopicPartitionKey,
+	row *model.RowChangedEvent,
+) error {
+	table, ok := b.activeTables.Load(row.Table.TableID)
+	if !ok {
+		tb := newTableStatus(key, row)
+		b.activeTables.Store(tb.id, tb)
+		// Send bootstrap message immediately when a new table is added
+		err := b.sendBootstrapMsg(ctx, tb)
```

Comment:
Will this sent the first bootstrap event ?

Reply:
Yes.